lib/propolis/src/enlightenment/hyperv/overlay.rs

```diff
@@ -0,0 +1,754 @@
+        };
+
+        // The overlay contents may be stored in the table itself (for pending
+        // overlays) or in guest memory (for active overlays). Before bothering
+        // to obtain them, create a mapping for the target page to verify that
+        // the target PFN is valid. This mapping will be used later if this is
+        // the first overlay for the target PFN.
+        let mut to_mapping = to_pfn.map(&memctx)?;
+
+        // Get the page contents to transfer, but don't actually change the
+        // tables yet, since applying the new overlay might still fail if there
+        // is an existing overlay of this kind at the target PFN.
+        //
+        // It's possible to do this check now, modify the tables in-place, and
+        // then assert that adding the new overlay succeeded, but this sequence
+        // allows this function to reuse the existing overlay removal code. This
+        // is unlikely to be a bottleneck in practice because guests change
+        // their overlay locations so infrequently (and any guest that spends
+        // all its time moving overlays around has, in some ways, asked for
+        // it...).
+        let contents = if from_set.active == kind {
+            let mut from_mapping = from_pfn
+                .map(&memctx)
+                .expect("active overlay PFNs can be mapped");
+            let mut contents = OverlayContents::default();
+            from_mapping.read_page(&mut contents);
+            contents
+        } else if let Some(pending_contents) = from_set.pending.get(&kind) {
+            pending_contents.clone()
+        } else {
+            return Err(OverlayError::NoOverlayOfKind(from_pfn.into(), kind));
+        };
+
+        self.add_overlay_using_mapping(&mut to_mapping, kind, contents)?;
+        self.remove_overlay(from_pfn, kind, acc_mem)
+            .expect("already found and read existing overlay");
+
+        Ok(())
+    }
+}
+
+/// An overlay tracker that adds, removes, and moves overlays.
+///
+/// # Usage requirements
+///
+/// After calling [`OverlayManager::new`], the caller must call
+/// [`OverlayManager::attach`] to attach the manager's memory accessor to a VM's
+/// memory hierarchy before it can add any overlays.
+///
+/// The [`OverlayPage`] structs returned from [`OverlayManager::add_overlay`]
+/// will attempt to remove themselves when they are dropped and assert that they
+/// can do so successfully. This will trigger a panic during VM shutdown if an
+/// overlay page with an active manager is dropped after the VM's memory context
+/// has been released. To avoid this, a Hyper-V stack that owns an overlay
+/// manager can drop its reference to the manager in its [`Lifecycle::halt`]
+/// callout; at that point no vCPUs are active or will become active, so this
+/// should remove the last strong reference to the manager, after which it is
+/// safe to drop the pages it produced.
+///
+/// [`Lifecycle::halt`]: crate::lifecycle::Lifecycle::halt
```

Comment:
This is the comment that I mentioned above that I would have really liked to encounter earlier in the file. Not a huge deal, though.

Reply:
Fixed in 094bd27.